### PR TITLE
Adds postprocessor to track time step number for transient problems

### LINKED ISSUE #29127: Add a postprocessor that returns the time step number
### Motivation

This is a minor quality-of-improvement update to allow users to easily track the time step count during a MOOSE simulation.

### Design

This information is already stored in `_feproblem.timeStep();`

### Impact

Users will be able to track the time step number using a ready-to-use postprocessor, instead of relying on, say, a `ParsedPostProcessor` using `TimePostprocessor` and `TimestepSize`.

### PR DESCRIPTION
**Pull request by anshchaube**
Minor quality-of-life improvement for users who want to track time step counts during their MOOSE simulations.

closes #29127

### CODE REVIEW

File: framework/doc/content/source/postprocessors/NumTimeSteps.md [hunk 1]
```diff
@@ -0,0 +1,17 @@
+# TimestepSize
```
Comment (GiudGiud) [inline framework/doc/content/source/postprocessors/NumTimeSteps.md:1]:
```suggestion
# NumTimeSteps
```

File: framework/doc/content/source/postprocessors/NumTimeSteps.md [hunk 2]
```diff
@@ -0,0 +1,17 @@
...
+
+!syntax description /Postprocessors/NumTimeSteps
+
+## Description
+
+`NumTimeSteps` reports the timestep size.
+
```
Comment (GiudGiud) [inline framework/doc/content/source/postprocessors/NumTimeSteps.md:8]:
```suggestion
```

File: framework/include/postprocessors/NumTimeSteps.h [hunk 3]
```diff
@@ -0,0 +1,31 @@
...
+//* https://github.com/idaholab/moose/blob/master/COPYRIGHT
+//*
+//* Licensed under LGPL 2.1, please see LICENSE for details
+//* https://www.gnu.org/licenses/lgpl-2.1.html
+
+#pragma once
+
+#include "GeneralPostprocessor.h"
+
+class NumTimeSteps : public GeneralPostprocessor
```
Comment (GiudGiud) [inline framework/include/postprocessors/NumTimeSteps.h:14]:
```suggestion
/**
 * Reports on the number of time steps already taken in the transient
 */
class NumTimeSteps : public GeneralPostprocessor
```

File: test/tests/postprocessors/num_time_steps/tests [hunk 4]
```diff
@@ -0,0 +1,9 @@
+[Tests]
+  [num_time_steps]
+    type = 'CSVDiff'
+    input = 'numtimesteps.i'
+    csvdiff = 'numtimesteps_out.csv'
+    requirement = 'The postprocessor `NumTimeSteps` shall count the number of time-steps accurately'
```
Comment (GiudGiud) [inline test/tests/postprocessors/num_time_steps/tests:6]:
```suggestion
    requirement = 'The shall be able to count the number of time-steps that have been taken '
```

File: test/tests/postprocessors/num_time_steps/tests [hunk 5]
```diff
@@ -0,0 +1,9 @@
...
+                   'over the course of a transient solve'
```
Comment (GiudGiud) [inline test/tests/postprocessors/num_time_steps/tests:7]:
```suggestion
                   'over the course of the simulation of a transient.'
```

**Review by GiudGiud — CHANGES_REQUESTED** [review]
nice this is almost ready

**Comment by GiudGiud** [comment]
let's squash all these commits togther in just a
```
address review
- updated wording
- added docstrings
```
commit message

File: test/tests/postprocessors/num_time_steps/moose.i [hunk 6]
```diff
@@ -0,0 +1,98 @@
+[Mesh]
```
Comment (loganharbour) [inline test/tests/postprocessors/num_time_steps/moose.i:1]: There is no need for a complex solve to test this. The test should be as minimal as possible to satisfy the requirement listed in the test spec:

> The shall be able to count the number of time-steps that have been taken over the course of the simulation of a transient

A coupled heat conduction solve and a large 27k element mesh is absolutely overkill for testing this postprocessor.

A basic problem without a solve is sufficient:

```
[Mesh/gmg]
  type = GeneratedMeshGenerator
  dim = 1
[]

[Postprocessors/timestep_ctr]
  type = NumTimeSteps
[]

[Problem]
  solve = False
[]

[Executioner]
  type = Transient
  dt = 1
  num_steps = 10
[]

[Outputs]
  csv = true
[]
```

**Review by loganharbour — DISMISSED** [review]

File: test/tests/postprocessors/num_time_steps/moose.i [hunk 7]
```diff
@@ -0,0 +1,98 @@
...
+    type = NumTimeSteps
+  []
+[]
+
+
+[Executioner]
+  type = Transient
+  nl_abs_tol = 1e-10
+  nl_rel_tol = 1e-4
+  dt = 1
```
Comment (dschwen) [inline test/tests/postprocessors/num_time_steps/moose.i:89]: I suggest setting `dt` to something that doesn't result the time being the exact same as the time step number...

**Review by GiudGiud — APPROVED** [review]

File: test/tests/postprocessors/num_time_steps/tests [hunk 8]
```diff
@@ -0,0 +1,11 @@
+[Tests]
+  issues = #29127
```
Comment (GiudGiud) [inline test/tests/postprocessors/num_time_steps/tests:2]:
```suggestion
  issues = '#29127'
```

**Comment by GiudGiud** [comment]
Test failures unrelated (fenix failures are known, unrelated tests w/ openmpi)

**Pull request merged**